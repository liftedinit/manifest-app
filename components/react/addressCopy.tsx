import React, { useState, useEffect } from 'react';
import { FiCopy, FiCheck } from 'react-icons/fi';

export const TruncatedAddressWithCopy = ({
  address,
  slice,
  size,
}: {
  address: string;
  slice: number;
  size?: string;
}) => {
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    if (copied) {
      timer = setTimeout(() => setCopied(false), 2000);
    }
    return () => clearTimeout(timer);
  }, [copied]);

  const handleCopy = async (e: React.MouseEvent<HTMLDivElement>) => {
    e.stopPropagation();
    e.preventDefault();
    try {
      await navigator.clipboard.writeText(address);
      setCopied(true);
    } catch (err) {
      console.error('Failed to copy: ', err);
    }
  };

  const truncatedAddress = `${address?.slice(0, slice)}...${address?.slice(-6)}`;
  const iconSize = size === 'small' ? 10 : 16;

  return (
    <span
      className="flex items-center hover:text-primary dark:hover:text-primary space-x-2 text-[#00000099] dark:text-[#FFFFFF99]"
      onClick={handleCopy}
      style={{ cursor: 'pointer' }}
    >
      <span className="truncate ">{truncatedAddress}</span>
      {copied ? <FiCheck size={iconSize} /> : <FiCopy size={iconSize} />}
    </span>
  );
};

export const AddressWithCopy = ({ address }: { address: string }) => {
  const [copied, setCopied] = useState(false);

  useEffect(() => {
    let timer: ReturnType<typeof setTimeout>;
    if (copied) {
      timer = setTimeout(() => setCopied(false), 2000);
    }
    return () => clearTimeout(timer);
  }, [copied]);

  const handleCopy = async () => {
    try {
      await navigator.clipboard.writeText(address);
      setCopied(true);
    } catch (err) {
      console.error('Failed to copy: ', err);
    }
  };

  return (
    <span
      className="flex items-center space-x-2"
      onClick={handleCopy}
      style={{ cursor: 'pointer' }}
    >
      <span className="hover:text-primary">{address}</span>
      {copied ? <FiCheck size="16" /> : <FiCopy size="16" />}
    </span>
  );
};
