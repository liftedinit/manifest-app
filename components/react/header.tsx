import env from '@/config/env';
import { WalletSection } from '../wallet';

export default function Header() {
  return (
    <header>
      <nav className="bg-white border-b border-gray-200 dark:bg-gray-800 dark:border-gray-700 px-4 fixed w-full lg:px-6 py-2 ">
        <div className="flex justify-end items-center mx-auto max-w-screen-6xl">
          {/* Log Out Button */}
          <WalletSection chainName={env.chain} />
        </div>
      </nav>
    </header>
  );
}
