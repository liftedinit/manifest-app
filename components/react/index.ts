export * from './addressCopy';
export * from './modal';
export * from './views';
export * from './settingsModal';
export * from './inputs';
export { SignModal } from './authSignerModal';
export * from './ModalDialog';
