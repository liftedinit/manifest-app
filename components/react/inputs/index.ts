export * from './AmountInput';
export * from './TextInput';
export * from './NumberInput';
export * from './TextArea';
