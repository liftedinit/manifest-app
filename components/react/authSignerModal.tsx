import { useEffect, useState } from 'react';
import { SignData } from '@cosmos-kit/web3auth';
import { TxBody, AuthInfo } from '@liftedinit/manifestjs/dist/codegen/cosmos/tx/v1beta1/tx';
import { decodePubkey } from '@cosmjs/proto-signing';
import { useWallet, useChain } from '@cosmos-kit/react';
import { getRealLogo } from '@/utils';
import { useTheme } from '@/contexts';
import env from '@/config/env';
import { ArrowRightIcon } from '../icons';
import { objectSyntax } from '../messageSyntax';
import { MsgSend } from '@liftedinit/manifestjs/dist/codegen/cosmos/bank/v1beta1/tx';
import {
  MsgCreateGroupWithPolicy,
  MsgSubmitProposal,
  MsgUpdateGroupMembers,
  MsgUpdateGroupPolicyMetadata,
  MsgUpdateGroupPolicyDecisionPolicy,
  MsgUpdateGroupMetadata,
} from '@liftedinit/manifestjs/dist/codegen/cosmos/group/v1/tx';
import {
  MsgCancelUpgrade,
  MsgSoftwareUpgrade,
} from '@liftedinit/manifestjs/dist/codegen/cosmos/upgrade/v1beta1/tx';
import { MsgSetPower } from '@liftedinit/manifestjs/dist/codegen/strangelove_ventures/poa/v1/tx';
import {
  MsgPayout,
  MsgBurnHeldBalance,
} from '@liftedinit/manifestjs/dist/codegen/liftedinit/manifest/v1/tx';
import {
  MsgSetDenomMetadata,
  MsgCreateDenom,
} from '@liftedinit/manifestjs/dist/codegen/osmosis/tokenfactory/v1beta1/tx';

type DisplayDataToSignProps = {
  data: SignData;
  address: string;
};

// Message decoder registry
const messageDecoders: Record<string, (value: Uint8Array) => any> = {
  '/cosmos.bank.v1beta1.MsgSend': (value: Uint8Array) => {
    const decoded = MsgSend.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgCreateGroupWithPolicy': (value: Uint8Array) => {
    const decoded = MsgCreateGroupWithPolicy.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgSubmitProposal': (value: Uint8Array) => {
    const decoded = MsgSubmitProposal.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupMetadata': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupMetadata.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupPolicyMetadata': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupPolicyMetadata.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupPolicyDecisionPolicy': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupPolicyDecisionPolicy.decode(value);
    return { ...decoded };
  },
  '/cosmos.group.v1.MsgUpdateGroupMembers': (value: Uint8Array) => {
    const decoded = MsgUpdateGroupMembers.decode(value);
    return { ...decoded };
  },
  '/cosmos.upgrade.v1beta1.MsgCancelUpgrade': (value: Uint8Array) => {
    const decoded = MsgCancelUpgrade.decode(value);
    return { ...decoded };
  },
  '/cosmos.upgrade.v1beta1.MsgSoftwareUpgrade': (value: Uint8Array) => {
    const decoded = MsgSoftwareUpgrade.decode(value);
    return { ...decoded };
  },
  '/strangelove_ventures.poa.v1.MsgSetPower': (value: Uint8Array) => {
    const decoded = MsgSetPower.decode(value);
    return { ...decoded };
  },
  '/liftedinit.manifest.v1.MsgPayout': (value: Uint8Array) => {
    const decoded = MsgPayout.decode(value);
    return { ...decoded };
  },
  '/liftedinit.manifest.v1.MsgBurnHeldBalance': (value: Uint8Array) => {
    const decoded = MsgBurnHeldBalance.decode(value);
    return { ...decoded };
  },
  '/osmosis.tokenfactory.v1beta1.MsgSetDenomMetadata': (value: Uint8Array) => {
    const decoded = MsgSetDenomMetadata.decode(value);
    return { ...decoded };
  },
  '/osmosis.tokenfactory.v1beta1.MsgCreateDenom': (value: Uint8Array) => {
    const decoded = MsgCreateDenom.decode(value);
    return { ...decoded };
  },
};

const DisplayDataToSign = ({
  data,
  address,
  className,
  addressClassName,
  txInfoClassName,
  theme,
}: DisplayDataToSignProps & {
  className?: string;
  addressClassName?: string;
  txInfoClassName?: string;
  theme?: string;
}) => {
  const [isAddressExpanded, setIsAddressExpanded] = useState(false);
  const [isTxInfoExpanded, setIsTxInfoExpanded] = useState(false);

  const decodeBodyBytes = (bodyBytes: Uint8Array) => {
    try {
      const decodedBody = TxBody.decode(bodyBytes);
      return {
        messages: decodedBody.messages.map(msg => {
          const base64Value = Buffer.from(msg.value).toString('base64');

          try {
            // Check if we have a specific decoder for this message type
            if (messageDecoders[msg.typeUrl]) {
              return {
                typeUrl: msg.typeUrl,
                value: messageDecoders[msg.typeUrl](msg.value),
              };
            }

            // Fallback to generic base64 decoding
            const decodedValue = Buffer.from(base64Value, 'base64').toString('utf8');
            try {
              return {
                typeUrl: msg.typeUrl,
                value: JSON.parse(decodedValue),
              };
            } catch {
              return {
                typeUrl: msg.typeUrl,
                value: decodedValue,
              };
            }
          } catch (error) {
            console.error(`Failed to decode message of type ${msg.typeUrl}:`, error);
            return {
              typeUrl: msg.typeUrl,
              value: base64Value,
            };
          }
        }),
        memo: decodedBody.memo,
        timeoutHeight: decodedBody.timeoutHeight.toString(),
        extensionOptions: decodedBody.extensionOptions,
        nonCriticalExtensionOptions: decodedBody.nonCriticalExtensionOptions,
      };
    } catch (error) {
      console.error('Failed to decode bodyBytes:', error);
      return 'Failed to decode bodyBytes';
    }
  };

  const decodeAuthInfoBytes = (authInfoBytes: Uint8Array) => {
    try {
      const decodedAuthInfo = AuthInfo.decode(authInfoBytes);
      return {
        signerInfos: decodedAuthInfo.signerInfos.map(signerInfo => ({
          publicKey: signerInfo.publicKey ? decodePubkey(signerInfo.publicKey) : null,
          modeInfo: signerInfo.modeInfo,
          sequence: signerInfo.sequence.toString(),
        })),
        fee: {
          amount: decodedAuthInfo.fee?.amount,
          gasLimit: decodedAuthInfo.fee?.gasLimit.toString(),
          payer: decodedAuthInfo.fee?.payer,
          granter: decodedAuthInfo.fee?.granter,
        },
      };
    } catch (error) {
      console.error('Failed to decode authInfoBytes:', error);
      return 'Failed to decode authInfoBytes';
    }
  };

  const formatValue = (value: any, theme: string) => {
    if (value instanceof Uint8Array) {
      return Buffer.from(value).toString('base64');
    }
    if (typeof value === 'object' && value !== null) {
      if ('bodyBytes' in value && 'authInfoBytes' in value) {
        const decodedValue = {
          ...value,
          bodyBytes: decodeBodyBytes(value.bodyBytes),
          authInfoBytes: decodeAuthInfoBytes(value.authInfoBytes),
        };
        return objectSyntax(
          JSON.parse(
            JSON.stringify(decodedValue, (_, v) => (typeof v === 'bigint' ? v.toString() : v))
          ),
          theme
        );
      }
      return objectSyntax(
        JSON.parse(JSON.stringify(value, (_, v) => (typeof v === 'bigint' ? v.toString() : v))),
        theme
      );
    }
    return String(value);
  };

  return (
    <div className={className}>
      <div className="flex flex-col gap-2">
        <button
          onClick={() => setIsAddressExpanded(!isAddressExpanded)}
          className="flex items-center gap-2 text-sm font-medium"
        >
          <div className="flex items-center gap-2 flex-row justify-between">
            <span>Address</span>
            {isAddressExpanded ? (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(90deg)' }} />
            ) : (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(180deg)' }} />
            )}
          </div>
        </button>
        {isAddressExpanded && <pre className={addressClassName}>{address}</pre>}
      </div>
      <div className="flex flex-col gap-2">
        <button
          onClick={() => setIsTxInfoExpanded(!isTxInfoExpanded)}
          className="flex items-center gap-2 text-sm font-medium"
        >
          <div className="flex items-center gap-2 flex-row justify-between">
            <span>Tx Info</span>
            {isTxInfoExpanded ? (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(90deg)' }} />
            ) : (
              <ArrowRightIcon size={12} style={{ transform: 'rotate(180deg)' }} />
            )}
          </div>
        </button>
        {isTxInfoExpanded && (
          <div className={txInfoClassName}>{formatValue(data.value, theme ?? 'light')}</div>
        )}
      </div>
    </div>
  );
};

const SignModal = ({
  visible,
  onClose,
  data,
  approve,
  reject,
}: {
  visible: boolean;
  onClose: () => void;
  data: SignData;
  approve: () => void;
  reject: () => void;
}) => {
  const wallet = useWallet();
  const { address } = useChain(env.chain);
  const walletIcon = wallet.wallet?.logo;
  const walletName = wallet.wallet?.prettyName;
  const { theme } = useTheme();

  useEffect(() => {
    const modal = document.getElementById('sign-modal') as HTMLDialogElement;
    if (visible) {
      modal.showModal();
    } else {
      modal.close();
    }
  }, [visible]);

  const walletIconString = walletIcon?.toString() ?? '';

  return (
    <dialog id="sign-modal" className="modal top-0 right-0">
      <div className="modal-box max-w-lg w-full dark:bg-[#1D192D] bg-[#FFFFFF] rounded-lg shadow-xl">
        <div className="flex justify-between items-center pb-4">
          <div className="flex items-center gap-3">
            <img
              src={getRealLogo(walletIconString, theme === 'dark')}
              alt="Wallet type logo"
              className="w-8 h-8"
            />
            <h3 className="text-xl font-semibold">{walletName?.toString()} Direct Signer</h3>
          </div>
          <button className="btn btn-sm btn-circle btn-ghost" onClick={onClose}>
            ✕
          </button>
        </div>

        <DisplayDataToSign
          data={data}
          address={address ?? ''}
          theme={theme}
          className="space-y-4"
          addressClassName="p-3 rounded-md text-sm overflow-auto h-12 dark:bg-[#E0E0FF0A] bg-[#E0E0FF0A] dark:border-[#FFFFFF33] border-[#00000033] border"
          txInfoClassName="p-3 rounded-md text-sm overflow-auto h-[32rem] dark:bg-[#E0E0FF0A] bg-[#E0E0FF0A] dark:border-[#FFFFFF33] border-[#00000033] border"
        />

        <div className="modal-action mt-6 flex justify-between gap-4">
          <button
            className="btn flex-1 rounded-[12px] focus:outline-none dark:bg-[#FFFFFF0F] bg-[#0000000A]"
            onClick={() => {
              reject();
              onClose();
            }}
          >
            Reject
          </button>
          <button
            className="btn btn-gradient flex-1 rounded-[12px]"
            onClick={() => {
              approve();
              onClose();
            }}
          >
            Approve
          </button>
        </div>
      </div>
      <form method="dialog" className="modal-backdrop">
        <button onClick={onClose}>close</button>
      </form>
    </dialog>
  );
};

export default SignModal;
