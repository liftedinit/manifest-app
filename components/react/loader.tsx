import { useEffect, useState } from 'react';

interface LoaderProps {
  progress: number;
  loaded: boolean;
  onEnter: () => void;
}

export const Loader = ({ progress, loaded, onEnter }: LoaderProps) => {
  const [showEnterButton, setShowEnterButton] = useState(false);

  useEffect(() => {
    if (loaded) {
      // Add a small delay before showing the enter button
      const timer = setTimeout(() => setShowEnterButton(true), 500);
      return () => clearTimeout(timer);
    }
  }, [loaded]);

  return (
    <div className="flex flex-col items-center">
      <progress className="progress progress-primary w-56" value={progress} max="100" />
      <div className="mt-2 text-sm text-gray-500">Loading... {Math.round(progress)}%</div>
      {showEnterButton && (
        <button
          onClick={onEnter}
          className="mt-8 px-8 py-3 bg-primary text-white rounded-md opacity-0 animate-fade-in hover:bg-primary-dark transition-colors duration-300"
        >
          Enter
        </button>
      )}
    </div>
  );
};
