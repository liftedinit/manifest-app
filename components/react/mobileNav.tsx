import { useTheme } from '@/contexts/theme';
import Image from 'next/image';
import Link from 'next/link';
import { TailwindModal } from './modal';

import {
  GroupsIcon,
  BankIcon,
  FactoryIcon,
  AdminsIcon,
  LightIcon,
  DarkIcon,
  ArrowRightIcon,
} from '@/components/icons';
import { WalletSection } from '../wallet';
import { RiMenuUnfoldFill } from 'react-icons/ri';
import { useState } from 'react';
import { MdOutlineNetworkPing, MdContacts } from 'react-icons/md';

export default function MobileNav() {
  const closeDrawer = () => {
    const drawer = document.getElementById('my-drawer') as HTMLInputElement;
    if (drawer) drawer.checked = false;
  };

  const NavItem: React.FC<{ Icon: React.ElementType; href: string }> = ({ Icon, href }) => {
    return (
      <li>
        <Link href={href} legacyBehavior>
          <div
            onClick={closeDrawer}
            className="flex flex-row justify-start items-center transition-all duration-300 ease-in-out text-primary"
          >
            <Icon className="w-8 h-8" />
            <span className="text-2xl">{href.slice(1, 12)}</span>
          </div>
        </Link>
      </li>
    );
  };

  const { toggleTheme, theme } = useTheme();
  const [isDark, setIsDark] = useState(theme === 'dark');
  const [isContactsOpen, setContactsOpen] = useState(false);

  return (
    <>
      <div className="fixed top-0 left-0 right-0 z-40 p-3 bg-base-300 flex lg:hidden flex-row justify-between items-center">
        <Image src="/logo.svg" height={38} width={38} alt="manifest" />
        <label htmlFor="my-drawer" className="btn btn-sm btn-primary drawer-button">
          <RiMenuUnfoldFill fontSize={'24px'} />
        </label>
      </div>
      <div className="drawer z-50 lg:hidden">
        <input id="my-drawer" type="checkbox" className="drawer-toggle" />

        <div className="drawer-side min-h-screen h-full ">
          <label htmlFor="my-drawer" aria-label="close sidebar" className="drawer-overlay"></label>
          <ul className="menu p-4 w-80 min-h-full bg-[#F4F4FF] dark:bg-[#1D192D] space-y-3 text-base-content flex flex-col">
            <div className="flex flex-row justify-between items-center">
              <div className="flex flex-row gap-4 justify-between items-center">
                <Image src={'/logo.svg'} alt="logo" width={42} height={42} />
                <span className="text-2xl leading-tight text-balance">Alberto</span>
              </div>

              {/* Updated Theme Toggle */}
              <label className="swap swap-rotate text-[#00000066] dark:text-[#FFFFFF66] hover:text-primary dark:hover:text-primary transition-all duration-300 ease-in-out">
                <input
                  type="checkbox"
                  className="theme-controller hidden"
                  value="light"
                  checked={isDark}
                  onChange={() => {
                    setIsDark(!isDark);
                    toggleTheme();
                  }}
                />
                <DarkIcon className="swap-on fill-current w-9 h-9 duration-300" />
                <LightIcon className="swap-off fill-current w-9 h-9 duration-300" />
              </label>
            </div>
            <div className="divider divider-horizon"></div>
            <NavItem Icon={BankIcon} href="/bank" />
            <NavItem Icon={GroupsIcon} href="/groups" />
            <NavItem Icon={AdminsIcon} href="/admins" />
            <NavItem Icon={FactoryIcon} href="/factory" />

            <div className="divider divider-horizon"></div>

            {/* Added Endpoint Selector and Contacts buttons */}
            <li>
              <button
                onClick={() => {
                  const modal = document.getElementById(
                    'endpoint_selector_modal'
                  ) as HTMLDialogElement;
                  if (modal) modal.showModal();
                }}
                className="flex flex-row justify-start items-center transition-all duration-300 ease-in-out text-primary"
              >
                <MdOutlineNetworkPing className="w-8 h-8" />
                <span className="text-2xl">Endpoints</span>
              </button>
            </li>
            <li className="mb-4">
              <button
                onClick={() => {
                  setContactsOpen(true);
                }}
                className="flex flex-row justify-start items-center transition-all duration-300 ease-in-out text-primary"
              >
                <MdContacts className="w-8 h-8" />
                <span className="text-2xl">Contacts</span>
              </button>
            </li>

            <div className="justify-between items-center">
              <WalletSection chainName="manifest" />
            </div>

            {/* Updated close button - now uses flex-1 and mt-auto to push to bottom */}
            <div className="flex justify-start mt-auto">
              <button onClick={closeDrawer} className="btn btn-sm btn-outline btn-primary">
                <ArrowRightIcon fontSize={'24px'} />
              </button>
            </div>
          </ul>
        </div>
      </div>
      <TailwindModal
        isOpen={isContactsOpen}
        setOpen={setContactsOpen}
        showContacts={true}
        onSelect={undefined}
      />
    </>
  );
}
