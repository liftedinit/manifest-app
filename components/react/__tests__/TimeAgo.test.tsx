import { cleanup, render } from '@testing-library/react';
import { afterEach, describe, expect, jest, spyOn, test } from 'bun:test';

import { TimeAgo } from '@/components';
import { formatComponent } from '@/tests';

describe('TimeAgo', () => {
  afterEach(() => {
    cleanup();
    jest.restoreAllMocks();
  });

  test('renders correctly in the past', () => {
    const now = new Date('2020-01-10 00:00:00 UTC');

    let mockup = render(
      <TimeAgo datetime={new Date('2020-01-09 00:00:00 UTC')} relativeDate={now} />
    );
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('1-day');
    expect(mockup.getByText(/1 day ago/i)).toBeInTheDocument();

    mockup = render(<TimeAgo datetime={new Date('2020-01-05 00:00:00 UTC')} relativeDate={now} />);
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('5-days');
    expect(mockup.getByText(/5 days ago/i)).toBeInTheDocument();

    mockup = render(<TimeAgo datetime={new Date('2020-01-01 00:00:00 UTC')} relativeDate={now} />);
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('1-week');
    expect(mockup.getByText(/1 week ago/i)).toBeInTheDocument();
  });

  test('renders correctly in the future', () => {
    const now = new Date('2020-01-01 00:00:00 UTC');

    let mockup = render(
      <TimeAgo datetime={new Date('2020-01-09 00:00:00 UTC')} relativeDate={now} />
    );
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('1-week');
    expect(mockup.getByText(/in 1 week/i)).toBeInTheDocument();

    mockup = render(<TimeAgo datetime={new Date('2020-01-05 00:00:00 UTC')} relativeDate={now} />);
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('4-days');
    expect(mockup.getByText(/in 4 days/i)).toBeInTheDocument();

    mockup = render(<TimeAgo datetime={new Date('2020-01-02 00:00:00 UTC')} relativeDate={now} />);
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('1-day');
    expect(mockup.getByText(/in 1 day/i)).toBeInTheDocument();

    mockup = render(<TimeAgo datetime={new Date('2020-01-01 00:00:00 UTC')} relativeDate={now} />);
    expect(formatComponent(mockup.asFragment())).toMatchSnapshot('just-now');
    expect(mockup.getByText(/just now/i)).toBeInTheDocument();
  });

  test('updates live', () => {
    jest.setSystemTime(new Date('2020-01-09 00:00:00 UTC'));
    const setInterval$ = spyOn(global, 'setInterval');

    const comp = <TimeAgo datetime={new Date('2020-01-10 00:00:00 UTC')} />;
    const mockup = render(comp);
    expect(mockup.getByText(/in 1 day/i)).toBeInTheDocument();

    expect(setInterval$).toHaveBeenCalledTimes(1);

    jest.setSystemTime(new Date('2020-01-09 23:00:00 UTC'));
    setInterval$.mock.calls.forEach(([callback]) => {
      callback();
    });

    mockup.rerender(comp);
    expect(mockup.getByText(/in 1 hour/i)).toBeInTheDocument();
  });
});
