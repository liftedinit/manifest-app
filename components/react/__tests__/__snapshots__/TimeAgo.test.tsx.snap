// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAgo renders correctly in the past: 1-day 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 9, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-09T00:00:00.000Z"
  >
    1 day ago
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the past: 5-days 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 5, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-05T00:00:00.000Z"
  >
    5 days ago
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the past: 1-week 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 1, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-01T00:00:00.000Z"
  >
    1 week ago
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the future: 1-week 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 9, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-09T00:00:00.000Z"
  >
    in 1 week
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the future: 4-days 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 5, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-05T00:00:00.000Z"
  >
    in 4 days
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the future: 1-day 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 2, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-02T00:00:00.000Z"
  >
    in 1 day
  </time>
</DocumentFragment>"
`;

exports[`TimeAgo renders correctly in the future: just-now 1`] = `
"<DocumentFragment>
  <time
    class="tooltip tooltip-primary "
    data-tip="January 1, 2020 at 12:00:00 AM UTC"
    datetime="2020-01-01T00:00:00.000Z"
  >
    just now
  </time>
</DocumentFragment>"
`;
